feat: add controlled shutdown methods

Popax21 opened the pull request:
See prior discussion in #1041. This PR adds an `(async) fn shutdown(&self) -> Result<()>` method to the `(Async)Transport` traits which can be used to shut down the transport in a controlled manner. Currently, these are only implemented for the SMTP transports when the pool feature is enabled, since otherwise there's nothing to shut down. For these transports, `send_raw` will return an error after the transport has been shut down, however other transports will continue to function after a shutdown, since their shutdowns are simple noops (this could be changed to be more consistent by adding a shutdown flag to all transports, but I've decided not to do this for the initial PR).

paolobarbolini commented on src/transport/smtp/pool/async_impl.rs:
```diff
@@ -134,10 +146,36 @@ impl<E: Executor> Pool<E> {
     }
 
+    pub async fn shutdown(&self) -> Result<(), Error> {
+        if let Some(connections) = self.connections.lock().await.take() {
```
This doesn't unlock the mutex until the end of the function, so anyone trying to get a connection will wait for shutdown to finish instead of immediately getting an error.

See the print order in: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=370b2dc1bdaf98b485d2ba591930f7fa

Popax21 replied:
Addressed as part of the rebase in 13d7982.

paolobarbolini commented on src/transport/smtp/pool/async_impl.rs:
```diff
@@ -134,10 +146,36 @@ impl<E: Executor> Pool<E> {
     }
 
+    pub async fn shutdown(&self) -> Result<(), Error> {
+        if let Some(connections) = self.connections.lock().await.take() {
```
This would also be a good place to shutdown `handle`.

Popax21 replied:
I agree; however, with the current setup this would require moving the handle from an `OnceLock` to a full-blown `Mutex`, which I have decided not to implement initially for the same reason as not shutting down the sync thread - the task will immediately exit upon waking up, and pretty much consumes no resources while in this "zombie" state. If this is a worthwhile tradeoff for you though then I'll implement this change as part of the next round of feedback.

paolobarbolini replied:
Ah right another instance of `async-std` ruining things for us :roll_eyes:. tokio's `JoinHandle::abort` takes '&self`, async-std `JoinHandle::cancel` takes `self`. We could use [`futures_util::future::AbortHandle`](https://docs.rs/futures-util/0.3.31/futures_util/future/struct.AbortHandle.html) to get the `&self` API for both.

Popax21 replied:
Resolved in latest push

paolobarbolini reviewed (changes requested):
Thanks for the PR. I've left some comments on the async APIs which also apply to the sync side.

This is a good improvement but for the `sync` API it doesn't solve the fact that the thread won't be shutdown until `thread::sleep` finishes sleeping.

What we could do is:

1. replace `thread::sleep` with [`thread::park_timeout`](https://doc.rust-lang.org/1.85.0/std/thread/fn.park_timeout.html), have the implementation handle spurious unparks, and then [`Thread::unpark`](https://doc.rust-lang.org/1.85.0/std/thread/struct.Thread.html#method.unpark) on `Drop`/`shutdown` of the transport.
2. use [`mpsc::Receiver::recv_timeout`](https://doc.rust-lang.org/1.85.0/std/sync/mpsc/struct.Receiver.html#method.recv_timeout) instead of `thread::sleep`, compared to the previous solution have no spurious unparks, and just close the `mpsc::Sender` on the `Drop`/`shutdown` side.

The second is a bit simpler and could also allow future implementations to send messages to the thread before closing the channel.

paolobarbolini commented:
Sorry I caused some merge conflicts with #1058. I've changed from `pub` to `pub(crate)`

Popax21 commented:
> This is a good improvement but for the `sync` API it doesn't solve the fact that the thread won't be shutdown until `thread::sleep` finishes sleeping.

I was considering implementing such a capability in the shutdown method, but decided against it - the thread will exit immediately after waking up, making it practically consume no resources while it is in this "zombie" state, while prematurely waking up the thread would require additional implementation complexity. However, looking into it some more I think a `thread::park_timeout` based solution like the one you mentioned should be simple enough in that regard - I'll investigate implementing it once I get around to addressing your feedback.

Popax21 commented on src/transport/smtp/error.rs:
```diff
@@ -139,6 +146,7 @@ impl fmt::Display for Error {
             Kind::Transient(code) => {
                 write!(f, "transient error ({code})")?;
             }
+            Kind::TransportShutdown => f.write_str("transport has been shut down")?,
```
Addressed as part of the rebase in 13d7982.

Popax21 commented on src/transport/smtp/pool/async_impl.rs:
```diff
@@ -134,10 +146,36 @@ impl<E: Executor> Pool<E> {
+                })
+                .await;
+
+            res.take().unwrap_or(Ok(()))
```
Do you propose that the `shutdown` method should not return a Result? Currently I've opted for it to do so, since this way the user gets informed about anything potentially going wrong as part of the shutdown (which for me includes a broken connection), but if you prefer that such errors are quietly "swallowed" then this logic can be removed.

Popax21 replied:
Resolved in latest push

paolobarbolini reviewed (approved):
LGTM. Thank you.
